src/session_pool/session_pool.js

```diff
@@ -0,0 +1,153 @@
+import EventEmitter from 'events';
+import { openKeyValueStore } from '../key_value_store';
+import Session from './session';
+import events from '../events';
+import { ACTOR_EVENT_NAMES_EX } from '../constants';
+
+
+/**
+ * Handles the session rotation, creation and persistence.
+ *
+ */
+
+// TODO: We should probably add some debug loging
+// TODO: Validation
+export default class SessionPool extends EventEmitter {
+    /**
+     *
+     * @param [options]
+     * @param options.initialPoolSize {Number}
+     * @param options.maxPoolSize {Number}
+     * @param options.maxSessionAgeSecs {Number}
+     * @param options.maxSessionReuseCount {Number}
+     * @param options.persistStateKeyValueStoreId {String}
+     */
+    constructor({
+        maxPoolSize = 1000,
+        maxSessionAgeSecs = 3000,
+        maxSessionReuseCount = 50,
+
+        persistStateKeyValueStoreId = null,
+        persistStateKey = 'SESSION_POOL_STATE',
+
+        createSessionFunction = null,
+
+    }) {
+        super();
+
+        // Pool Configuration
+        this.maxPoolSize = maxPoolSize;
+        this.createSessionFunction = createSessionFunction || this._defaultCreateSessionFunction;
+
+        // Session configuration
+        // @TODO: Maybe options.sessionOptions / this.sessionOptions?
+        this.maxSessionAgeSecs = maxSessionAgeSecs;
+        this.maxSessionReuseCount = maxSessionReuseCount;
+
+        // Session storage
+        this.persistStateKeyValueStoreId = persistStateKeyValueStoreId;
+        this.persistStateKey = persistStateKey;
+
+        // Statistics
+        this.activeSessions = 0;
+        this.blockedSessions = 0;
+
+        // Operative states
+        this.storage = null;
```

Comment:
why `this.storage` and not `this.keyValueStore` ?  `storage` is just a generic phrase, when looking at the code, you don't really see what that means..

Reply:
I wanted to use more generic name in case of using different storage than the KV store, but I guess it is not needed.

Reply:
Yeah, no need to optimize for future, better to have it clear now. We can change it if needed